proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,258 @@
+value `10`, upper bound `13`, space `0` and array index of `2`:
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 10, i32 13, i32 0, i8 1 }, i32 2, i1 false),
+```
+
+Initialization of other resources in the same array would differ only by the
+index. The values for upper and lower bound and register space will stay the
+same.
+
+https://godbolt.org/z/ff3bnKxov
+
+#### Fixed-size Multi-dimensional Array
+
+Groups of resources can also be declared as multi-dimensional arrays of fixed
+size
+
+```
+RWBuffer<float> B[4][4] : register(u2);
+RWBuffer<float> C[2][2][5] : register(u10, space1);
+```
+
+When initializing individual resources of multi-dimensional array the arguments
+of `@dx.op.createHandleFromBinding` look like as if the array was flattened to a
+single dimension.
+
+That means the array `B[4][4]` is treated as if it was declared as `B[16]`, and
+`C[2][2][5]` as if it was `C[20]`.
+
+For example a call to initialize `C[1][0][3][0]` from the array declared above
+would have lower bound value `10`, upper bound `29`, space `1`, and array index
+`23` = `10 + 1*2*5 + 0*5 + 3`:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 10, i32 29, i32 1, i8 1 }, i32 23, i1 false), 
+```
+
+https://godbolt.org/z/8MMWheeGc
+
+#### Arrays of Unbounded Size
+
+Groups of resource arrays can be declared with undefined size. These are called
+_unbounded arrays_:
+```
+RWBuffer<float> D[] : register(u5);
+```
+For unbounded arrays the `@dx.op.createHandleFromBinding` calls to initialize
+individual resource handles will have upper bound set to `-1`. For example an
+initialization of handle `D[100]` from the `D` array declared above looks like
+this:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 105, i1 false)
+```
+https://godbolt.org/z/hx99s5dzP
+
+#### Unbounded-size Arrays and Multiple Dimensions
+
+For multi-dimensional arrays of resources only the first dimension can be
+unbounded:
+```
+RWBuffer<float> E[][4] : register(u5);
+```
+When initializing individual resources of an unbounded multi-dimensional array,
+the arguments of `@dx.op.createHandleFromBinding` again look like as if the
+array was flattened to a single dimension. For example the call to create
+resource handle `E[100][4]` from array `E` declared above will have index `409`
+= `5 + 4 x 100 + 4`:
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 409, i1 false),
+```
+https://godbolt.org/z/YrEnYhrEj
+
+Arrays with unbounded sizes in the other dimensions are not allowed, though the
+error message DXC reports does not clearly indicate this:
+
+```
+RWBuffer<float> F[4][];
+
+error: array has incomplete element type 'RWBuffer<float> []'
+```
+https://godbolt.org/z/osq7dzvdh
+
+### Resource Arrays at Local Scope
```

Comment:
I don't know how exhaustive you're trying to be here, but you can also put a resource array (or multi-dimensional resource array) inside a struct.

Reply:
Yeah, array of resources in struct and array of struct containing resource(s) should probably be called out, unless we want a separate document for that?